### ISSUE TITLE
Exploring Internal API improvements to Microstates

### ISSUE THREAD

**Issue by taras** [cowboyd/microstates.js#22]
We should be able to create an ObservableMicrostate from a Microstate. 

For example,
```js
let ms = microstates(State);
let { transitions, subscribe ) = new MicrostateObservable(ms);
```

Instead, we're passing State into MicrostateObservable constructor. The constructor is repeating code from `microstates()` because it's difficult to wrap every transition in a function that would allow to `next` the value returned by the transition. In MicrostateObservable, `mapTransitions(tee, callback)` receives a callback that next the value.  It's also important to point out that the transitions tree, unlike state tree, is stable and doesn't change.

There might be another way to think about the relationship between state and transitions. We could think about transitions as a stream of transitions.

For example, we could do something like this,

```js
let { state, transitions } = microstates(State);

Observable.from(transitions).subscribe({
  next([ transition, ...args ]) {
    let newState = transition(...args);
    // at this point, we would dispatch an action in Redux or set state without Redux.
    // we would even push it into a stream of states if we have one
    // this will also make it possible to dispatch an action with the new state
  }
})
```

We still need a better way to map transitions to make it easier to wrap them with a callback. I'm open to suggestions on how we might do this.

**Comment by cowboyd** [comment]
The more and more I think about it, creating the microstate before creating the observable is not a good API. I think you suggested it once, why not just say:

```
// microstate
let { state, transitions }  = microstates(Type, initialValue);

// observable
let { state, subscribe, transitions } = MicrostateObservable(Type, initialValue)
```

Because I see what you're saying, in the `ms` you're bound to a particular value, but an observable will represent a stream of values, so to instantiate it from a thing associated with a single value almost makes no sense.

The other thing that I was thinking, what if you think of the transitions themselves as a function that can take a value first, and then is curried with a particular value.

```
let transitions = transitionsFor(Boolean);
transitions(false).toggle() //=> true
```

or, alternatively, we could have the transitions be a tree with uncurried transition functions:

```
let transitions = new Transitions(Boolean);
transitions.toggle(false) //=> true

class User {
  name = String
  email = String
}

let transitions = new Transitions(User);
transitions.email.set({name: 'charles', email: 'cowboyd@frontside.io'}, 'cowboyd@frontside.io')

**Comment by cowboyd** [comment]
Another possibility to consider is that you have a fully generic transition function, and that we implement microstates in layers by successively currying in more context. Consider this: 

```js
function transition(fn, lens, root, currentValue, ...args) {
   //make the magic happen.
}
```

I can now set username on a user like:

```
transition((b) => b, lensPath(['name']), {name: 'charles', email: 'cowboyd@frontside.io'}, 'Fred');
//=> {name: 'Fred', email: 'cowboyd@frontside.io'}
```

When we build the transitions tree, we just curry in the transition fn and lens:

```
setName = transitions(b) => b, lensPath(['name']);

**Comment by taras** [comment]
Resolved by #25
